src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -755,2514 +345,391 @@ MulticopterPositionControl::poll_subscriptions()
-		/* we have a direction change */
-		intention = direction_change;
+				if (_in_smooth_takeoff) {
+					constraints.speed_up = _takeoff_speed;
+					// during smooth takeoff we disable yaw command
+					setpoint.yaw = setpoint.yawspeed = NAN;
+				}
+			}
 
-	} else {
-		/* catchall: acceleration */
-		intention = acceleration;
-	}
+			if (_vehicle_land_detected.landed && !_in_smooth_takeoff && !PX4_ISFINITE(setpoint.thrust[2])) {
+				// Keep throttle low
+				setpoint.thrust[0] = setpoint.thrust[1] = setpoint.thrust[2] = 0.0f;
+				setpoint.x = setpoint.y = setpoint.z = NAN;
+				setpoint.vx = setpoint.vy = setpoint.vz = NAN;
+				setpoint.yawspeed = NAN;
+				setpoint.yaw = _states.yaw;
+				constraints.landing_gear = vehicle_constraints_s::GEAR_KEEP;
+			}
 
+			// limit altitude only if local position is valid
+			if (PX4_ISFINITE(_states.position(2))) {limit_altitude(setpoint);};
 
-	/*
-	 * update user intention
-	 */
+			// Update states, setpoints and constraints.
+			_control.updateConstraints(constraints);
 
-	/* we always want to break starting with slow deceleration */
-	if ((_user_intention_xy != brake) && (intention  == brake)) {
+			_control.updateState(_states);
 
-		if (_jerk_hor_max.get() > _jerk_hor_min.get()) {
-			_manual_jerk_limit_xy = (_jerk_hor_max.get() - _jerk_hor_min.get()) / _velocity_hor_manual.get() *
-						sqrtf(_vel(0) * _vel(0) + _vel(1) * _vel(1)) + _jerk_hor_min.get();
+			_control.updateSetpoint(setpoint);
 
-			/* we start braking with lowest accleration */
-			_acceleration_state_dependent_xy = _deceleration_hor_slow.get();
+			// Generate desired thrust and yaw.
+			_control.generateThrustYawSetpoint(_dt);
 
-		} else {
+			matrix::Vector3f thr_sp = _control.getThrustSetpoint();
 
-			/* set the jerk limit large since we don't know it better*/
-			_manual_jerk_limit_xy = 1000000.f;
+			// Adjust thrust setpoint based on landdetector only if the
+			// vehicle is NOT in pure Manual mode.
+			if (!_in_smooth_takeoff && !PX4_ISFINITE(setpoint.thrust[2])) {limit_thrust_during_landing(thr_sp);}
 
-			/* at brake we use max acceleration */
-			_acceleration_state_dependent_xy = _acceleration_hor_max.get();
+			// Fill local position, velocity and thrust setpoint.
+			_local_pos_sp.timestamp = hrt_absolute_time();
+			_local_pos_sp.x = _control.getPosSp()(0);
+			_local_pos_sp.y = _control.getPosSp()(1);
+			_local_pos_sp.z = _control.getPosSp()(2);
+			_local_pos_sp.yaw = _control.getYawSetpoint();
+			_local_pos_sp.yawspeed = _control.getYawspeedSetpoint();
 
-		}
+			_local_pos_sp.vx = _control.getVelSp()(0);
+			_local_pos_sp.vy = _control.getVelSp()(1);
+			_local_pos_sp.vz = _control.getVelSp()(2);
+			thr_sp.copyTo(_local_pos_sp.thrust);
 
-		/* reset slew rate */
-		_vel_sp_prev(0) = _vel(0);
-		_vel_sp_prev(1) = _vel(1);
+			// Fill attitude setpoint. Attitude is computed from yaw and thrust setpoint.
+			_att_sp = ControlMath::thrustToAttitude(thr_sp, _control.getYawSetpoint());
```

Comment:
`thrustToAttitude()` holds its own copy of the complete attitude setpoint message. That overhead can be saved by reseting the outside attitude setpoint message and passing a reference to the function such that it can fill the fields.